ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLogWorker.java

```diff
@@ -208,8 +243,7 @@ synchronized void updateIndex(long i) {
     this.workerThread = new Thread(this::run, name);
 
     // Server Id can be null in unit tests
-    metricRegistry.addDataQueueSizeGauge(queue);
-    metricRegistry.addLogWorkerQueueSizeGauge(writeTasks.q);
+    metricRegistry.addQueueSizeGauges(queues);
     metricRegistry.addFlushBatchSizeGauge(() -> (Gauge<Integer>) () -> flushBatchSize);
```

Comment:
The `syncBatchSize` gauge this creates still holds `SegmentedRaftLogWorker` after `close()`.

Reply:
@adoroszlai ,  good catch.  Just have changed it to use AtomicInteger.

Reply:
Unfortunately the lambda still has a reference to the enclosing `SegmentedRaftLogWorker` instance.

Reply:
The following POC resolves this specific leak (please feel free to adapt):

```diff
diff --git ratis-server/src/main/java/org/apache/ratis/server/metrics/SegmentedRaftLogMetrics.java ratis-server/src/main/java/org/apache/ratis/server/metrics/SegmentedRaftLogMetrics.java
index 50d582ac..d213563b 100644
--- ratis-server/src/main/java/org/apache/ratis/server/metrics/SegmentedRaftLogMetrics.java
+++ ratis-server/src/main/java/org/apache/ratis/server/metrics/SegmentedRaftLogMetrics.java
@@ -24,6 +24,7 @@ import com.codahale.metrics.Timer;
 import org.apache.ratis.protocol.RaftGroupMemberId;
 import org.apache.ratis.server.raftlog.segmented.SegmentedRaftLogCache;
 
+import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.ToIntFunction;
 
@@ -113,8 +114,10 @@ public class SegmentedRaftLogMetrics extends RaftLogMetricsBase {
     });
   }
 
-  public void addFlushBatchSizeGauge(MetricRegistry.MetricSupplier<Gauge> supplier) {
-    registry.gauge(RAFT_LOG_SYNC_BATCH_SIZE, supplier);
+  public AtomicInteger addFlushBatchSizeGauge() {
+    MutableIntegerGaugeSupplier migs = new MutableIntegerGaugeSupplier();
+    registry.gauge(RAFT_LOG_SYNC_BATCH_SIZE, migs);
+    return migs.asAtomicInteger();
   }
 
   private Timer getTimer(String timerName) {
@@ -181,4 +184,23 @@ public class SegmentedRaftLogMetrics extends RaftLogMetricsBase {
   public void onStateMachineDataReadTimeout() {
     registry.counter(RAFT_LOG_STATEMACHINE_DATA_READ_TIMEOUT_COUNT).inc();
   }
+
+  public static class MutableIntegerGaugeSupplier implements MetricRegistry.MetricSupplier<Gauge>, Gauge<Integer> {
+    private final AtomicInteger value = new AtomicInteger();
+
+    public AtomicInteger asAtomicInteger() {
+      return value;
+    }
+
+    @Override
+    public Gauge<Integer> newMetric() {
+      return this;
+    }
+
+    @Override
+    public Integer getValue() {
+      return value.get();
+    }
+  }
+
 }
diff --git ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLogWorker.java ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLogWorker.java
index a5b12eb9..fa4516e1 100644
--- ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLogWorker.java
+++ ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLogWorker.java
@@ -211,7 +211,7 @@ class SegmentedRaftLogWorker {
   private final long segmentMaxSize;
   private final long preallocatedSize;
   private final RaftServer.Division server;
-  private final AtomicInteger flushBatchSize = new AtomicInteger();
+  private final AtomicInteger flushBatchSize;
 
   private final boolean asyncFlush;
   private final boolean unsafeFlush;
@@ -243,7 +243,7 @@ class SegmentedRaftLogWorker {
 
     // Server Id can be null in unit tests
     metricRegistry.addQueueSizeGauges(queues);
-    metricRegistry.addFlushBatchSizeGauge(() -> flushBatchSize::get);
+    flushBatchSize = metricRegistry.addFlushBatchSizeGauge();
     this.logFlushTimer = metricRegistry.getFlushTimer();
     this.raftLogSyncTimer = metricRegistry.getRaftLogSyncTimer();
     this.raftLogQueueingTimer = metricRegistry.getRaftLogQueueTimer(
```

With this, we still have a leak due to `lastLeaderElapsedTime` gauge in `LeaderElectionMetrics` (at least).

Reply:
@adoroszlai , thanks for catching it.  I mean to pass the AtomicInteger and use lambda in SegmentedRaftLogMetrics.  Then, it won't have a reference to SegmentedRaftLogWorker.  Let me try it.